### PR TITLE
[codecs] initial type definitions

### PR DESCRIPTION
**Pull request by martinheidegger**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@martinheidegger thx!

File: types/codecs/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+/// <reference types="node" />
+
+declare namespace codecs {
+    type JsonObject = { [Key in string]?: JsonValue };
+    interface JsonArray extends Array<JsonValue> {}
+    type JsonValue = string | number | boolean | null | JsonObject | JsonArray;
+
+    interface BaseCodec<InType = any, OutType = InType> {
+        encode(input: InType): Buffer;
+        decode(input: Uint8Array): OutType;
```
Comment (amcasey) [inline types/codecs/index.d.ts:15]: I'm probably missing something obvious, but I'm not sure why the parameter type isn't `Buffer`.

**Review by amcasey — APPROVED** [review]

File: types/codecs/index.d.ts (hunk 1 above)
Comment (martinheidegger) [reply types/codecs/index.d.ts:15]: The project guarantees to return a Buffer instance when encoding but accepts a Uint8Array when decoding.

**Comment by martinheidegger** [comment]
Ready to merge

**Pull request merged**